### PR TITLE
feat(install): Allow downloading from private repositories

### LINKED ISSUE #4243: Using Scoop for private tools
I've searched around but have been unable to find a clear answer to this question. Can Scoop be configured to download artifacts from private (authenticated) URLs, such as a Github Release from a private repository)? Thanks!

### PR DESCRIPTION
**Pull request by tmsmith**
Closes #4243. 

Github Private repositories return a 404 when trying to use the releases download even with a provided API Token. To solve this I used the releases API which returns an asset URL that will allow for the download request. Also with this PR, a section can be defined in the config file to inject any custom headers needed for other buckets.

```
{
    "lastupdate":  "2021-02-09T20:48:57.2632262-05:00",
    "hosts":  [
                  {
                      "match":  "https://localhost/repos.*",
                      "headers":  "@{Authorization=token ABC}"
                  },
                  {
                      "match":  "https://somewhere.com/.*",
                      "headers":  "@{Authorization=token XXX}"
                  }
              ],
    "checkver_token":  "<GH Token>"
}
```

### CODE REVIEW

**Comment by timharris777** [comment]
Waiting for this feature!

**Comment by JustinLoyed** [comment]
Any update if this can be reviewed and included?

**Comment by astelmachonak-nydig** [comment]
This is great feature - waiting for it to start using scoop as "homebrew for Windows" at my company!

**Comment by rashil2000** [comment]
What scopes are required in the GH token used here? Is it a plain token? 

We already use a GH API token for making authenticated requests, we can use it here too:
https://github.com/ScoopInstaller/Scoop/blob/7d5a47cc774961c15fdae6be1daf6331282b0161/libexec/scoop-config.ps1#L79-L82

**Comment by astelmachonak-nydig** [comment]
@rashil2000 I think it is `repo` scope, probably same as required for `checkver_token` token, so we definitely can merge this two. However name `checkver_token` will be confusing, should we rename it to `github_token` and use for both cases?

**Comment by rashil2000** [comment]
Hmm, yes, renaming to `github_token` or `gh_api_token` should do

**Comment by rashil2000** [comment]
@tmsmith can you update this PR with latest develop?

**Comment by astelmachonak-nydig** [comment]
It looks like URL syntax `<URL>#/filename` is broken for the private repos, investigating right now.

**Comment by astelmachonak-nydig** [comment]
I have a fix for that. If nobody objects, I will create a new PR with github private thing only (custom headers can be added later - I don't have setup to test them). 
@rashil2000 should I rename `checkver_token` to `gh_api_token` in this PR?

**Comment by astelmachonak-nydig** [comment]
@tmsmith I have a validated fix for issue described [https://github.com/ScoopInstaller/Scoop/pull/4254#issuecomment-1065282843](here). How do you want me to share it? Should I make a PR to https://github.com/tmsmith/scoop/tree/private-github-support or just share a code snippet with you?


Here it is, changes are in line with regex and line where `url` is set.
```powershell
    # Github.com
    if ($url -match "github.com\/(?<owner>[^\/]+)\/(?<repo>[^\/]+)\/releases\/download\/(?<tag>[^\/]+)\/(?<file>[^\/#]+)(?<filename>.*)" -and (get_config 'gh_api_token')) {
        $headers = @{
            "Authorization" = "token $(get_config 'gh_api_token')"
        }
        $privateUrl = "https://api.github.com/repos/$($matches['owner'])/$($matches['repo'])"
        $assetUrl="https://api.github.com/repos/$($matches['owner'])/$($matches['repo'])/releases/tags/$($matches['tag'])"

        $isPrivate = (Invoke-RestMethod -Uri $privateUrl -Headers $headers).private
        if ($isPrivate) {
            $url = ((Invoke-RestMethod -Uri $assetUrl -Headers $headers).assets | Where-Object { $_.name -eq $matches['file'] }).url + $($matches['filename'])
        }
    }
```

**Comment by tmsmith** [comment]
PR is fine.

**Comment by astelmachonak-nydig** [comment]
@tmsmith please check my comment above and fix the issue with using '#/` in private repositories URLs

**Comment by tmsmith** [comment]
I updated the fork with the code snippet you posted above.

**Comment by astelmachonak-nydig** [comment]
@tmsmith Thanks! I re-checked the current PR code and it works fine for me with `gh_api_token` set. @rashil2000 I think it is good to go now. There is definitely a need to combine/rework `SCOOP_CHECKVER_TOKEN`, `checkver-token` and `gh_api_token`, so we have single environment variable and single configuration property which work same across the code base, however this can be done in separate PR (I plan to open one once this one is merged). Do you have any concerns about this approach?

**Comment by rashil2000** [comment]
Actually, let's keep using `checkver_token` in this PR too. We can consolidate the name changes properly in a separate PR.

@tmsmith please update the changelog.

**Comment by astelmachonak-nydig** [comment]
@tmsmith will you be able to update the PR as requested? I really want this feature to be in place before I start rolling out scoop as internal tool for the company...

**Comment by tmsmith** [comment]
I have update the files as requested. I currently do not have a windows machine that functions so I was unable to fully test the changes.

**Comment by astelmachonak-nydig** [comment]
@tmsmith I can do the test. However can you please use `checkver_token` instead of `gh_api_token`? This is what @rashil2000 asked, so we reuse existing tokens instead of adding new. Thanks!

**Comment by astelmachonak-nydig** [comment]
@tmsmith I did some testing for current PR state and while github stuff works fine, the proposed solution of custom headers doesn't work. I spent some time fixing it and here is what I got:

This is how hosts-related section of `dl` function is `install.ps1` looks like:
```powershell
        $hosts = get_config "hosts"
        if ($hosts) {
            $hosts | Where-Object { $url -match $_.match } | ForEach-Object {
                (ConvertFrom-StringData -StringData $_.headers).GetEnumerator() | ForEach-Object {
                    $wreq.headers[$_.Key] = $_.Value
                }
            }
        }
```

And here is example of configuration file hosts section definition:

```json
    "hosts": [
        {
            "match": ".*api.github.com/repos.*",
            "headers": "Authorization=token SECRET_TOKEN\nCustomHeader=Custom Header Value"
        }
    ]
```
You can see: headers are defined using `KEY=VALUE` syntax, multiple values can be specified using `\n` (the JSON newline character). This setup is validated and works - header(s) get populated as expected.

Hopefully with this change/testing done, we can finally get this merged! Thanks @tmsmith and @rashil2000!

**Comment by tmsmith** [comment]
updated to match your changes and use `checkver_token`.

**Comment by astelmachonak-nydig** [comment]
Thank you, @tmsmith!

@rashil2000 I can confirm this code works as expected, hopefully we can approve/merge now!

File: lib/core.ps1 [hunk 1]
```diff
@@ -1030,6 +1030,21 @@ function handle_special_urls($url)
         # Reshapes the URL to avoid redirections
         $url = "https://downloads.sourceforge.net/project/$($matches['project'])/$($matches['file'])"
     }
+
+    # Github.com
+    if ($url -match "github.com\/(?<owner>[^\/]+)\/(?<repo>[^\/]+)\/releases\/download\/(?<tag>[^\/]+)\/(?<file>[^\/#]+)(?<filename>.*)" -and (get_config 'gh_api_token')) {
```
Comment (rashil2000) [inline lib/core.ps1:1035]:
```suggestion
    if ($url -match "github.com\/(?<owner>[^\/]+)\/(?<repo>[^\/]+)\/releases\/download\/(?<tag>[^\/]+)\/(?<file>[^\/#]+)(?<filename>.*)" -and (get_config 'checkver_token')) {
```

File: lib/core.ps1 [hunk 2]
```diff
@@ -1030,6 +1030,21 @@ function handle_special_urls($url)
...
+        $headers = @{
+            "Authorization" = "token $(get_config 'gh_api_token')"
```
Comment (rashil2000) [inline lib/core.ps1:1037]:
```suggestion
            "Authorization" = "token $(get_config 'checkver_token')"
```

**Comment by astelmachonak-nydig** [comment]
Uups! This changes were in place, but looks like were lost after force-push. I will re-test branch after fixes from @tmsmith

**Comment by astelmachonak-nydig** [comment]
Retested with latest changes - everything looks good

**Review by rashil2000 — DISMISSED** [review]

File: lib/core.ps1 [hunk 3]
```diff
@@ -1030,6 +1030,21 @@ function handle_special_urls($url)
         # Reshapes the URL to avoid redirections
         $url = "https://downloads.sourceforge.net/project/$($matches['project'])/$($matches['file'])"
     }
+
+    # Github.com
+    if ($url -match "github.com\/(?<owner>[^\/]+)\/(?<repo>[^\/]+)\/releases\/download\/(?<tag>[^\/]+)\/(?<file>[^\/#]+)(?<filename>.*)" -and (get_config 'checkver_token')) {
```
Comment (niheaven) [inline lib/core.ps1:1035]:
```suggestion
    if ($url -match 'github.com/(?<owner>[^/]+)/(?<repo>[^/]+)/releases/download/(?<tag>[^/]+)/(?<file>[^/#]+)(?<filename>.*)' -and ($token = get_config 'checkver_token')) {
```

`/` needs not be escaped.

File: lib/core.ps1 [hunk 4]
```diff
@@ -1030,6 +1030,21 @@ function handle_special_urls($url)
...
+        $headers = @{
+            "Authorization" = "token $(get_config 'checkver_token')"
```
Comment (niheaven) [inline lib/core.ps1:1037]:
```suggestion
            "Authorization" = "token $token"
```

File: lib/core.ps1 [hunk 5]
```diff
@@ -1030,6 +1030,21 @@ function handle_special_urls($url)
...
+        }
+        $privateUrl = "https://api.github.com/repos/$($matches['owner'])/$($matches['repo'])"
+        $assetUrl="https://api.github.com/repos/$($matches['owner'])/$($matches['repo'])/releases/tags/$($matches['tag'])"
```
Comment (niheaven) [inline lib/core.ps1:1040]:
```suggestion
        $privateUrl = "https://api.github.com/repos/$($Matches.owner)/$($Matches.repo)"
        $assetUrl = "https://api.github.com/repos/$($Matches.owner)/$($Matches.repo)/releases/tags/$($Matches.tag)"
```

File: lib/core.ps1 [hunk 6]
```diff
@@ -1030,6 +1030,21 @@ function handle_special_urls($url)
...
+
+        $isPrivate = (Invoke-RestMethod -Uri $privateUrl -Headers $headers).private
+        if ($isPrivate) {
+            $url = ((Invoke-RestMethod -Uri $assetUrl -Headers $headers).assets | Where-Object { $_.name -eq $matches['file'] }).url + $($matches['filename'])
```
Comment (niheaven) [inline lib/core.ps1:1044]:
```suggestion
        if ((Invoke-RestMethod -Uri $privateUrl -Headers $headers).Private) {
            $url = ((Invoke-RestMethod -Uri $assetUrl -Headers $headers).Assets | Where-Object -Property Name -EQ -Value $Matches.file).Url, $Matches.filename -join ''
```

File: lib/install.ps1 [hunk 7]
```diff
@@ -394,9 +394,22 @@ function dl($url, $to, $cookies, $progress) {
         if (-not ($url -imatch "sourceforge\.net" -or $url -imatch "portableapps\.com")) {
             $wreq.referer = strip_filename $url
         }
+        if ($url -imatch "api\.github\.com\/repos") {
```
Comment (niheaven) [inline lib/install.ps1:397]:
```suggestion
        if ($url -imatch "api\.github\.com/repos") {
```

**Review by niheaven — COMMENTED** [review]
Some static review, and please move `Features` to top of `Unrelease` in CHANGELOG.

Since I'm not using private repo, please check if the review comments works.

**Comment by astelmachonak-nydig** [comment]
Thanks @niheaven - I checked you changes with both github and private host repo and they work fine!

@tmsmith can you please incorporate proposed changes and rename `hosts` to `private_hosts` in code. This should be last change needed before merge. Thanks in advance for helping getting this merged - appreciate it!

**Comment by niheaven** [comment]
I've done some tweaks and rename config entry to `private_hosts`, and please check for the commits @tmsmith @astelmachonak-nydig @rashil2000. 

If no bugs, it could be merged, and if there're some that introduced by my commits, please revert them or fixed.

**Comment by astelmachonak-nydig** [comment]
I will re-test this tomorrow.

**Comment by astelmachonak-nydig** [comment]
Just tested - everything works as expected from current branch.

**Review by niheaven — DISMISSED** [review]

**Comment by astelmachonak-nydig** [comment]
Thanks @tmsmith @rashil2000 @niheaven 
This is great feature to have. When do you think it will be in master as part of next release?

**Comment by astelmachonak-nydig** [comment]
@rashil2000 looks like your last commit broke it when `private_hosts` variable is not set :( - I am getting error:
`Cannot bind argument to parameter 'StringData' because it is null.`

Will need to open the PR to fix it

**Comment by rashil2000** [comment]
> @rashil2000 looks like your last commit broke it when `private_hosts` variable is not set :( - I am getting error: `Cannot bind argument to parameter 'StringData' because it is null.`

Oh, sorry, my bad. I thought Where-Object and ForEach-Object functions would automatically take care of null input.

> Will need to open the PR to fix it

 Are you opening it?

**Comment by astelmachonak-nydig** [comment]
> Are you opening it?

Yes, shortly.

**Comment by astelmachonak-nydig** [comment]
@rashil2000 https://github.com/ScoopInstaller/Scoop/pull/4832

**Comment by niheaven** [comment]
> I thought Where-Object and ForEach-Object functions would automatically take care of null input.

Actually, they cannot, and that's why I rarely use them in pipeline.

**Comment by CEbbinghaus** [comment]
Is there documentation on this feature?

**Comment by rashil2000** [comment]
Run `scoop help config`

**Pull request merged**